File: backend/src/subscription/subscription.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../prisma.service';
import { SubscriptionStatus as PrismaSubscriptionStatus } from '@prisma/client';

export enum SubscriptionPlan {
  FREE = 'FREE',
  PREMIUM_MONTHLY = 'PREMIUM_MONTHLY',
  PREMIUM_YEARLY = 'PREMIUM_YEARLY',
}

// Re-export Prisma SubscriptionStatus for convenience
export const SubscriptionStatus = PrismaSubscriptionStatus;
export type SubscriptionStatus = PrismaSubscriptionStatus;

export interface SubscriptionCheckResult {
  hasAccess: boolean;
  subscription: {
    plan: string;
    status: string;
    expiresAt: Date | null;
  } | null;
}

@Injectable()
export class SubscriptionService {
  constructor(private prisma: PrismaService) {}

  /**
   * Check if user has active premium access
   * Admins and super admins always have premium access
   */
  async hasPremiumAccess(
    userId: string,
    userRole?: string,
  ): Promise<SubscriptionCheckResult> {
    // Admin and super admin users always have premium access
    if (userRole === 'ADMIN' || userRole === 'SUPER_ADMIN') {
      return {
        hasAccess: true,
        subscription: {
          plan: 'ADMIN',
          status: SubscriptionStatus.ACTIVE,
          expiresAt: null,
        },
      };
    }

    const subscription = await this.prisma.subscription.findFirst({
      where: {
        user_id: userId,
        status: 'ACTIVE',
      },
      orderBy: {
        created_at: 'desc',
      },
    });

    if (!subscription) {
      return {
        hasAccess: false,
        subscription: null,
      };
    }

    // Check if subscription has expired
    if (subscription.expires_at && subscription.expires_at < new Date()) {
      // Update status to expired
      await this.prisma.subscription.update({
        where: { id: subscription.id },
        data: { status: SubscriptionStatus.EXPIRED },
      });

      return {
        hasAccess: false,
        subscription: {
          plan: subscription.plan,
          status: SubscriptionStatus.EXPIRED,
          expiresAt: subscription.expires_at,
        },
      };
    }

    // Check if user has premium plan
    const hasPremium =
      subscription.plan === 'PREMIUM_MONTHLY' ||
      subscription.plan === 'PREMIUM_YEARLY';

    return {
      hasAccess: hasPremium,
      subscription: {
        plan: subscription.plan,
        status: subscription.status,
        expiresAt: subscription.expires_at,
      },
    };
  }

  /**
   * Get active subscription for a user
   */
  async getActiveSubscription(userId: string) {
    return this.prisma.subscription.findFirst({
      where: {
        user_id: userId,
        status: SubscriptionStatus.ACTIVE,
      },
      orderBy: {
        created_at: 'desc',
      },
    });
  }

  /**
   * Create a new subscription
   */
  async createSubscription(data: {
    userId: string;
    plan: SubscriptionPlan;
    status?: SubscriptionStatus;
    expiresAt?: Date;
  }) {
    return this.prisma.subscription.create({
      data: {
        user_id: data.userId,
        plan: data.plan,
        status: data.status || SubscriptionStatus.ACTIVE,
        expires_at: data.expiresAt,
      },
    });
  }

  /**
   * Update subscription status
   */
  async updateSubscription(
    id: string,
    data: {
      status?: SubscriptionStatus;
      expiresAt?: Date;
      cancelledAt?: Date;
    },
  ) {
    return this.prisma.subscription.update({
      where: { id },
      data: {
        status: data.status,
        expires_at: data.expiresAt,
        cancelled_at: data.cancelledAt,
      },
    });
  }

  /**
   * Cancel subscription
   */
  async cancelSubscription(id: string) {
    return this.prisma.subscription.update({
      where: { id },
      data: {
        status: SubscriptionStatus.CANCELLED,
        cancelled_at: new Date(),
      },
    });
  }
}

File: backend-spring/src/main/java/com/ulticode/modules/subscription/service/SubscriptionService.java
package com.ulticode.modules.subscription.service;

import com.ulticode.modules.subscription.dto.CreateSubscriptionDTO;
import com.ulticode.modules.subscription.dto.SubscriptionCheckResultDTO;
import com.ulticode.modules.subscription.dto.SubscriptionDTO;
import com.ulticode.modules.subscription.entity.Subscription;

import java.util.Optional;

/**
 * Service interface for subscription operations.
 */
public interface SubscriptionService {

    /**
     * Check if a user has active premium access.
     * Admins and super admins always have premium access.
     *
     * @param userId  the user ID
     * @param userRole the user's role (optional)
     * @return the check result with access status and subscription details
     */
    SubscriptionCheckResultDTO hasPremiumAccess(String userId, String userRole);

    /**
     * Check if the current authenticated user has premium access.
     * Admins and super admins always have premium access.
     *
     * @return true if the user has premium access
     */
    boolean hasPremiumAccess();

    /**
     * Get the active subscription for a user.
     *
     * @param userId the user ID
     * @return the active subscription or empty
     */
    Optional<Subscription> getActiveSubscription(String userId);

    /**
     * Get the current user's subscription status.
     *
     * @return the subscription DTO or null if no active subscription
     */
    SubscriptionDTO getCurrentUserSubscription();

    /**
     * Create a new subscription.
     *
     * @param dto    the subscription data
     * @param userId the user ID
     * @return the created subscription
     */
    SubscriptionDTO createSubscription(CreateSubscriptionDTO dto, String userId);

    /**
     * Update subscription status.
     *
     * @param id     the subscription ID
     * @param status the new status
     * @param userId the user ID (for validation)
     * @return the updated subscription
     */
    SubscriptionDTO updateSubscriptionStatus(String id, String status, String userId);

    /**
     * Cancel a subscription.
     *
     * @param id     the subscription ID
     * @param userId the user ID (for validation)
     * @return the cancelled subscription
     */
    SubscriptionDTO cancelSubscription(String id, String userId);

    /**
     * Find subscription by ID.
     *
     * @param id the subscription ID
     * @return the subscription or empty
     */
    Optional<Subscription> findById(String id);

    /**
     * Convert entity to DTO.
     *
     * @param subscription the entity
     * @return the DTO
     */
    SubscriptionDTO toDTO(Subscription subscription);
}
